---
title: "Case study: reanalysis of Cao et al. (2020)"
author: "Davide Risso"
date: "4/27/2021"
output: html_document
---

# Introduction

Cao et al. used sci-RNA-seq3 to create a human cell atlas of fetal gene expression.

The dataset is made of about 4 million cells.

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, error = FALSE, message = FALSE)
```

```{r packages}
library(zellkonverter)
library(scuttle)
library(scone)
library(DelayedArray)
library(scater)
library(mbkmeans)
library(HDF5Array)
```

```{r data_in}
dati <- "GSE156793_S3_gene_count.h5ad"
sce <- readH5AD(dati, use_hdf5 = TRUE)
assayNames(sce) <- "counts"
sce
```

# Normalization, filtering, PCA

```{r filtering}
num_reads = 1
num_cells = 0.01*ncol(sce)
system.time(is_exp = rowSums(assay(sce) >= num_reads ) >= num_cells)
table(is_exp)

sce <- sce[is_exp,]
```

```{r norm}
system.time(sce <- PsiNorm(sce))
sce <- logNormCounts(sce)
sce
```

```{r hvg}
vars <- rowVars(counts(sce))
names(vars) <- rownames(sce)
vars <- sort(vars, decreasing = TRUE)
sce <- sce[names(vars)[seq(1000)],]
sce
```

```{r realize}
saveHDF5SummarizedExperiment(sce, 
                             dir = "GSE156793_hvg", 
                             prefix="", replace=TRUE, 
                             level=NULL, verbose=FALSE)
```

```{r load}
sce <- loadHDF5SummarizedExperiment(dir = "GSE156793_hvg")
sce
```

```{r pca}
system.time(sce <- scater::runPCA(sce, 
                      BPPARAM = BiocParallel::MulticoreParam(5),
                      BSPARAM = BiocSingular::RandomParam()))
```

```{r clustering}
## here run mbkmeans
k_list <- seq(15, 30)
time.start <- proc.time()
km_res <- lapply(k_list, function(k) {
  mbkmeans(sce, clusters = k, 
           batch_size = 500,
           reduceMethod = "PCA",
           calc_wcss = TRUE,
           BPPARAM = MulticoreParam(nworkers)
  )
})
time.end <- proc.time()
time <- time.end - time.start
times[tt] <- time[3]
tt <- tt+1
cat("Clustering:", time[3]/60, "mins")

  saveRDS(km_res, file = here("main/case_studies/data/full/TENxBrainData",
                              "mbkmeans_pca50_k15-30_bs500.rds"))
  saveRDS(times, file = here("main/case_studies/data/full/TENxBrainData",
                             "computational_times.rds"))
} else {
  km_res <- readRDS(here("main/case_studies/data/full/TENxBrainData",
                         "mbkmeans_pca50_k15-30_bs500.rds"))
}
wcss <- sapply(km_res, function(x) sum(x$WCSS_per_cluster))
print(data.frame(k_list, wcss))
plot(k_list, wcss, type = "b")

clusters <- sapply(km_res, function(x) factor(paste0("mbk", x$Clusters)))
colnames(clusters) <- paste0("mbk_", k_list)
colData(sce) <- cbind(colData(sce), clusters)
```




```{r umap}

# scater::plotUMAP(sce, colour_by = "")
```
